### PR TITLE
Implemented  Authorization

### PR DESCRIPTION
**Pull request by dev-usmanarshad**
**Authorization:**
- Manager can do all
- Member of a Project can read the projects
- Member can create the task

**Other Modifications**
- Added name field in SignUp
- Worked on some styling

### CODE REVIEW

File: app/controllers/application_controller.rb [hunk 1]
```diff
@@ -1,2 +1,14 @@
...
+
+    rescue_from CanCan::AccessDenied do |exception|
+        redirect_to root_url, :alert => "You are un-authorize to perform this action."
+      end
+
+    protected
+
+    def configure_permitted_parameters
+      devise_parameter_sanitizer.permit(:sign_up, keys: [:name])
+    end
```
Comment (GS-Shahid) [inline app/controllers/application_controller.rb]: We need to improve spacing here

File: app/models/ability.rb [hunk 2]
```diff
@@ -0,0 +1,16 @@
+class Ability
+  include CanCan::Ability
+  def initialize(user)
+    user ||= User.new 
+    if user.role == "manager"
+      can :manage, :all
+    elsif user.role == "member"
+      can :read, Project
+      can [:read, :update], Task
+      can [:read, :update], Comment, user_id: user.id
```
Comment (GS-Shahid) [inline app/models/ability.rb:10]: member should be able to read all comments

File: app/models/ability.rb [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+    else
+      can :read, Project
+      can :read, Task
```
Comment (GS-Shahid) [inline app/models/ability.rb]: No need to worry or allow. We don't have guest users yet.

File: app/views/layouts/application.html.erb [hunk 4]
```diff
@@ -32,4 +32,4 @@
     <%= yield %>
 
       </body>
-    </html>
+    </html>
```
Comment (GS-Shahid) [inline app/views/layouts/application.html.erb]: Configure your editor to add a blank line at then end of every file

File: app/views/projects/_project.html.erb [hunk 5]
```diff
@@ -10,8 +10,11 @@
         <p class="card-text">Manager: <%= project.creator.name %></p>
         <p class="card-text">Members: <%= project.project_members.length %></p>
         <%= link_to "View", project_path(project), class: "btn btn-outline-success" %>
-        <% if user_signed_in? %>
+
+        <% if can? :update, project %>
           <%= link_to "Edit", edit_project_path(project), class: "btn btn-outline-info" %>
+        <% end %>
+        <% if can? :destory, project %>
```
Comment (GS-Shahid) [inline app/views/projects/_project.html.erb:17]: Seems like we also need to add authorization on controller actions.

File: app/models/ability.rb (hunk 2 above)
Comment (dev-usmanarshad) [reply app/models/ability.rb:10]: this is working as we expect. a member can read the other members comments. and member can edit his own comment.

**Pull request merged**